### DISCUSSION TITLE
How render the wide line?

### DISCUSSION THREAD

**Discussion by dotoritos-kim** [vagran/dxf-viewer#45] [Ideas]
Due to webgl constraints linesegment is limited to 1px.

I suggest [this ](https://github.com/spite/THREE.MeshLine)as a wide line rendering library.
If you think this is inappropriate, please share your thoughts and reasons.

**Comment by vagran** [comment]
The proper approach is to convert lines into mesh in scope of `DxfScene` class, using existing `TRIANGLES` batch type. Also note that line segment in DXF may have `bulge` attribute (some kind of curvature) and different start and end width, which also should be properly handled. Here is an example when both these properties are combined:
![image](https://user-images.githubusercontent.com/6065976/232047015-c4698450-5afb-4149-94a8-ab4846d98f5f.png)
Also the line may have styling applied (dashes) which also would be nice to implement for that code. In general, it is quite straight-forward feature, and even, as I remember, has proper placeholder in the code. I just was too lazy to implement it initially.

**Reply by dotoritos-kim** [reply]
Could you please share the dxf file of the example?

**Reply by vagran** [reply]
I have sent it privately to your e-mail, since this file unfortunately cannot be publicly shared.

**Reply by dotoritos-kim** [reply]
Haven't been sent yet. Did you send it?

**Reply by vagran** [reply]
Yes, I have sent. May be in spam folder?

**Reply by dotoritos-kim** [reply]
Sorry, please send again.
It didn't exist in the spam folder.

**Reply by vagran** [reply]
Done. Resent as link as well.

**Comment by dotoritos-kim** [comment]
I think we need to distinguish between constant width type and variable width type.
If you look at the [document](https://help.autodesk.com/view/OARX/2023/ENU/?guid=GUID-748FC305-F3F2-4F74-825A-61F04D757A50), it has data in the code 43 when it has a fixed width and 40, 41 when it has a variable width.

**Reply by vagran** [reply]
If we have a generic code which is able to render variable width, it could render constant width as well.

**Reply by dotoritos-kim** [reply]
Oops, I was stupid.

**Comment by dotoritos-kim** [comment]
To render variously, such as wideline, polyline, and line patterns, it seems that you will eventually need to write a shader.
I think this [repo](https://github.com/gl-vis/gl-line2d/tree/master/lib/shaders) will help us.

**Reply by vagran** [reply]
> To render variously, such as wideline, polyline, and line patterns, it seems that you will eventually need to write a shader. I think this [repo](https://github.com/gl-vis/gl-line2d/tree/master/lib/shaders) will help us.

Actually, it is mostly not about shader, the line should "materialized" - i.e. mesh is generated by the line parameters. May be pattern support can be implemented in shader, however I'm not sure about that, it may be impossible to have precise border for a dashes, it will not be a straight line, since the line parameter is interpolated independently in each triangle the line segment quad consists from. This is how GPU graphics pipeline works with vertex attributes. On the other hand it might be good enough. So this is will require some research work. And in any case first is line materialization without pattern support. 
![image](https://user-images.githubusercontent.com/6065976/235450097-c5b7930d-b1af-4a84-8598-31f56fd23f4b.png)

**Comment by dotoritos-kim** [comment]
![test](https://i.ibb.co/JtjfKKx/meshline.png)

I was able to carefully look at the properties of the meshlines and manipulate them to render the song plane in any shape I wanted. It seems to be available. Later, we'll see if dashlines can be rendered as well.
